### PR TITLE
Fetch key from JWKS URI if available

### PR DESCRIPTION
**Pull request by stanhu**
In non-standard OpenID Connect providers, such as Azure B2C, discovery
does not work because the discovery URL does not match the issuer field.
If a JWKS URI is provided when discovery is disabled, we should make an
HTTP request for the keys and use the response.

Closes https://github.com/m0n9oose/omniauth_openid_connect/issues/72

This is part of the effort to upstream changes in the GitLab fork: https://gitlab.com/gitlab-org/ruby/gems/gitlab-omniauth-openid-connect/-/issues/5.

### CODE REVIEW

**Comment by stanhu** [comment]
@jessieay @bufferoverflow Could you review this?

File: lib/omniauth/strategies/openid_connect.rb [hunk 1]
```diff
@@ -196,9 +196,13 @@ def authorize_uri
...
-        return config.jwks if options.discovery
-
-        key_or_secret || config.jwks
+        @public_key ||= if options.discovery
+                          config.jwks
+                        elsif key_or_secret
+                          key_or_secret
+                        elsif client_options.jwks_uri
+                          fetch_key
+                        end
```
Comment (jessieay) [inline lib/omniauth/strategies/openid_connect.rb:205]: **Question**: should there be an `else` option here that returns `config.jwks`?

Previous logic would return `config.jwks if options.discovery` but also fell back to `config.jwks` if `key_or_secret` were `nil`. 

In addition to adding `fetch_key`, we are changing the fallback/default behavior here, which could have unintended / backward incompatible consequences. 

What do you think?

**Review by jessieay — COMMENTED** [review]
One question for you @stanhu 

Also what do you think about adding docs on this / adding a note to CHANGELOG as part of this PR?

File: lib/omniauth/strategies/openid_connect.rb (hunk 1 above)
Comment (stanhu) [reply lib/omniauth/strategies/openid_connect.rb:205]: @jessieay It's a good question, but `config` always attempts to call out to the OpenID discovery URL, which sort of defeats the purpose if discovery is disabled.

**Comment by bufferoverflow** [comment]
@azure  Is this correct ? Azure B2C, discovery does not work because the discovery URL does not match the issuer field and therefore being non-standard OpenID Connect?

**Comment by stanhu** [comment]
@bufferoverflow FYI, the Azure non-compliance has been discussed in other issues:

1. https://github.com/MicrosoftDocs/azure-docs/issues/38427
1. https://github.com/coreos/go-oidc/issues/344

File: lib/omniauth/strategies/openid_connect.rb (hunk 1 above)
Comment (jessieay) [reply lib/omniauth/strategies/openid_connect.rb:205]: ah ok that makes sense @stanhu . In other words: the former default return value was problematic and what we are fixing 💯

**Review by jessieay — APPROVED** [review]
LGTM!

**Pull request merged**